Metrics

ZiadAbdelkarim opened the pull request:
From Joseph's original PR:

Created a base Metric class. Implemented two types of Metrics: Counts and Measures. Counts are values that increase monotonically and are tracked every round (e.g., moves, rounds, etc.) while Measures are system properties that are calculated less frequently, determinted by a fixed frequency. All Metrics are shown in the GUI.

Basic Counts (# Rounds, # Moves, and # Activations) have all been implemented as default.

Data can be output to .csv by using the "Data" button or with `Cmd`/`Ctrl` + M.

Resolves #2.

josephbriones committed:
Creation of base Metric class

+ implemented framework for the Metric class.

ZiadAbdelkarim committed:
Added Rounds and Movements classes within Metric.h

ZiadAbdelkarim committed:
metric.h file now contains "Measure" and "Count" classes

+Virtual classes Measure and Count and the corresponding variables and functions

josephbriones committed:
Implemented round, activation and movement counters

+ created Round, Movement and Activation Count classes.
~ program outputs each count per round

josephbriones committed:
Implemented metrics class

+ Metrics output to .CSV file
+ Data output triggered by 'Data' button click or by pressing 'M' on the keyboard

josephbriones committed:
Minor bug fix

-fixed misnaming of a signal

josephbriones committed:
Formatting & documentation fixes

+ added documentation for headers files

josephbriones committed:
Added relative path to file creation

+ dumps file to a user's home directory

josephbriones committed:
Added additional documentation

+ added additional comments to the header files

josephbriones committed:
Implemented metrics display

+displays information about all metrics with front-end ListView. 
~ replaced numRounds and numMovements

jdaymude commented:
Resolves issue #2.

jdaymude committed:
Rewrote/polished the metrics backend and updated documentation

~ Updated Count/Measure structure and the way AmoebotSystem maintains them
~ Updated metrics export to use a more detailed JSON structure
+ Added metrics documentation to Usage, including JSON structure
~ Polished interface and remapped some hotkeys
~ Removed extraneous comments and updated some header guards

jdaymude commented:
Updated export to use the JSON format (instead of CSV) and is performed with `Ctrl`/`Cmd + E`.

jdaymude committed:
Gave ScriptInterface access to exporting metrics and getting # objects

jdaymude committed:
Added a perimeter tracking to Compression to demonstrate measures

~ Refactored Measures to refine relationship with children
~ Updated metrics to use QString names instead of std::string
~ Fixed bug in how Simulator packages metrics for the GUI
~ Fixed memory leak in AmoebotSystem

jdaymude commented:
@josephbriones @ryanyiu This PR is now ready for review. The main contributions are:
- Adding `Count` and `Measure` classes in `core/metrics.*`.
- Replacing the hardcoded default metrics (# rounds and # moves) with three new counts: # Rounds, # Activations, and # Moves.
- Adding `PerimeterMeasure` to `alg/compression.*` to demonstrate how custom measures are created and linked to particle systems.
- Polishing the front-end GUI (fixing previous issues with overlapping text).
- Switching the data export format to JSON, modifying a few of the hotkeys, and updating the Usage documentation accordingly.
- Adding a new `ScriptInterface` command for exporting metrics JSON.

You can split up primary/secondary reviewing responsibilities however you'd like. Please be aware that there's one small issue we'll need to fix in the future:
- In the JSON data, I have a field `"algorithm"` that is supposed to identify the algorithm that was being tracked. However, our architecture is blocking me from getting this information easily, since only `ui/algorithm` knows the actual algorithm names. So currently, I have the JSON saying `"algorithm" : "???"`.

ryanyiu commented on core/amoebotsystem.h:
```diff
@@ -52,40 +53,44 @@ class AmoebotSystem : public System, public RandomNumberGenerator {
+
+  // Various access functions for metrics (counts and measures). getCounts
+  // (resp., getMeasures) returns a reference to the count (resp., measure)
+  // list. getCount (resp., getMeasure) return a reference to the named count
```
Small change: `returns` instead of `return`

jdaymude replied:
I've addressed this in a new commit.

ryanyiu commented on core/simulator.h:
```diff
@@ -31,22 +32,33 @@ class Simulator : public QObject {
+  // writing the metrics JSON to it.
+  void exportMetrics();
+
+  // TODO.
```
Did you want to keep the comment? Or should a GitHub issue be made for implementing this part (so the comment + function can be moved to a separate PR)?

jdaymude replied:
All the functions here were already implemented; it's just that `simulator.h` didn't have any header function comments. I've added one for `saveScreenshotSetup` in the most recent commit.

ryanyiu commented:
@jdaymude Everything looks good apart from one or two minor things, which I left comments on.

jdaymude committed:
Minor comment updates

josephbriones reviewed (commented):
It seems as though I am unable to generate the .json metrics file. Currently trying to see if this is a Mac issue.

Additionally, would you like all the copyrights updated to 2020?

ryanyiu reviewed (changes requested):
Can't seem to remove this comment, but everything was good (I got `json` files to generate on my end).

ryanyiu reviewed (dismissed):
LGTM

jdaymude commented on core/simulator.cpp:
```diff
@@ -78,12 +79,34 @@ int Simulator::numObjects() const {
+  }
+  QTextStream outStream(&outFile);
+  outStream << system->metricsAsJSON();
+}
```
@josephbriones No need to update the copyrights in this PR, though I should definitely do that soon. As for the JSON generation, not sure what would be going wrong. This is the relevant code; as you can see in lines 100-101, everything is being done in platform-independent Qt libraries. According to the documentation, `"/"` is always translated to the correct directory separator if a `QFile` is used.

jdaymude replied:
@josephbriones What happens when you try to generate the file? Does it crash? Or does it just not appear where you expect it to?

jdaymude replied:
After talking to Joseph, we found that a JSON file is indeed being written. However, `QDir::currentPath()` on macOS is pointing at `<builddir>/AmoebotSim.app/Contents/MacOS`, which is a hidden directory that isn't directly accessible by Finder. I'm investigating how we could patch this to point at a stable location on Windows and macOS so that users will know where to go look.

jdaymude replied:
Ok, I think I've found a solution I'm happy with. @josephbriones try replacing `QDir::currentPath()` with `QCoreApplication::applicationDirPath()`. This should be the `QString` path for the directory containing the executable, which is stable and visible on both Windows and macOS. If this works for you, I'll go ahead and push a new commit.

jdaymude committed:
Made the JSON file output more robust and consistent across platforms

ryanyiu reviewed (approved):
LGTM. As an aside, should we open up a separate GitHub issue for building some kind of notification to a user when a metric file is made? It feels a bit unresponsive to just press the button, and not see anything telling you that metric data has been created.

jdaymude commented:
@ryanyiu actually, that's exactly what happens when you use the `exportMetrics()` command on the command line. But emitting this signal from the button requires a lot of signal/slot piping that I thought was unnecessary at this point in time. Yet another thing to improve in the UI after we figure out #7 and #3...

josephbriones reviewed (approved):
LGTM.
